pandas/core/interval.py

```diff
@@ -0,0 +1,413 @@
+from pandas.core.common import (_values_from_object, _ensure_platform_int,
+                                _get_next_label, _get_prev_label)
+from pandas.core.index import (Index, _ensure_index, default_pprint,
+                               InvalidIndexError, MultiIndex)
+from pandas.lib import (Interval, IntervalMixin, IntervalTree,
+                        interval_bounds_to_intervals,
+                        intervals_to_interval_bounds)
+from pandas.util.decorators import cache_readonly
+import pandas.core.common as com
+
+
+_VALID_CLOSED = set(['left', 'right', 'both', 'neither'])
+
+
+def _get_interval_closed_bounds(interval):
+    """
+    Given an Interval or IntervalIndex, return the corresponding interval with
+    closed bounds.
+    """
+    left, right = interval.left, interval.right
+    if interval.open_left:
+        left = _get_next_label(left)
+    if interval.open_right:
+        right = _get_prev_label(right)
+    return left, right
+
+
+class IntervalIndex(IntervalMixin, Index):
+    _typ = 'intervalindex'
+    _comparables = ['name']
+    _attributes = ['name', 'closed']
+    _allow_index_ops = True
+    _engine = None # disable it
+
+    def __new__(cls, left, right, closed='right', name=None, fastpath=False):
+        # TODO: validation
+        result = IntervalMixin.__new__(cls)
+        result._left = _ensure_index(left)
+        result._right = _ensure_index(right)
+        result._closed = closed
+        result.name = name
+        if not fastpath:
+            result._validate()
+        result._reset_identity()
+        return result
+
+    def _validate(self):
+        # TODO: exclude periods?
+        if self.closed not in _VALID_CLOSED:
+            raise ValueError("invalid options for 'closed': %s" % self.closed)
+        if len(self.left) != len(self.right):
+            raise ValueError('left and right must have the same length')
+        if not (self.left <= self.right).all():
+            raise ValueError('left side of interval must be <= right side')
+
+    def _simple_new(cls, values, name=None, **kwargs):
+        # ensure we don't end up here (this is a superclass method)
+        raise NotImplementedError
+
+    def _cleanup(self):
+        pass
+
+    @property
+    def _engine(self):
+        raise NotImplementedError
+
+    @cache_readonly
+    def _tree(self):
+        return IntervalTree(self.left, self.right, closed=self.closed)
+
+    @property
+    def _constructor(self):
+        return type(self).from_intervals
+
+    @classmethod
+    def from_breaks(cls, breaks, closed='right', name=None):
+        return cls(breaks[:-1], breaks[1:], closed, name)
+
+    @classmethod
+    def from_intervals(cls, data, name=None):
+        data = np.asarray(data)
+        left, right, closed = intervals_to_interval_bounds(data)
+        return cls(left, right, closed, name)
+
+    @classmethod
+    def from_tuples(cls, data, closed='right', name=None):
```

Comment:
Yeah, I'm not sure we really need this constructor, though it is useful for constructing test cases rather than explicitly creating lots of Interval objects.